### PR TITLE
Fix JNLP launch on Windows.  Directory Create doesnt accept user and gro...

### LINKED ISSUE #310: Error "No mapping between account names and security IDs was done." when creating a new jenkins slave
Jenkins cookbook version: 2.2.1
Error message: Chef::Exceptions::Win32APIError: No mapping between account names and security IDs was done.
Already identified as https://github.com/opscode-cookbooks/jenkins/issues/241

I haven't seen that error when I used v1.2.X

Further details:

recipe [acme::slave]
```ruby
jenkins_windows_slave node['fqdn'] do
  description         'Runs a series of high-level smoke tests'
  remote_fs           "C:/jenkins"
  labels              ['windows']
  user                "build"
  password        "d3V4321"
  group              "Users"
  action              :create
	environment(
		'FOO' => 'bar',
		'BAZ' => 'qux',
	)
end
```

metadata
```ruby
name             'acme_jenkins'
version           '0.0.1'
depends        'jenkins', '= 2.2.1'
```

w7-001.acme.com.json
```json
{
  "run_list": [
    "recipe[acme_jenkins::slave]"
   ],
   "name": "w7-001.acme.com",
   "chef_environment": "test",
   "normal" : {
    "jenkins" : {
        "node" : {
            "agent_type"  : "windows",
            "description" : "IE 10 on Windows 7",
            "home"        : "D:/jenkins"
        },
        "windows" : {
            "service" : "interactive"
        }
    },
    "machine_attributes" : {
        "node"  : "build win7 clients"
    }
  }
}
```

```
[2015-01-06T13:01:45+00:00] INFO: Running queued delayed notifications before re-raising exception
[2015-01-06T13:01:45+00:00] ERROR: Running exception handlers
[2015-01-06T13:01:45+00:00] ERROR: Exception handlers complete
[2015-01-06T13:01:45+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
Chef Client failed. 1 resources updated
[2015-01-06T13:01:45+00:00] FATAL: Chef::Exceptions::Win32APIError: jenkins_windows_slave[w7-001.acme.com] (acme_jenkins::slave line 20) had an error: Chef::Exceptions::Win32APIError: directory[C:/jenkins] (dynamically defined) had an error: Chef::Exceptions::Win32APIError: No mapping between account names and security IDs was done.
---- Begin Win32 API output ----
System Error Code: 1332
System Error Message: No mapping between account names and security IDs was done.
---- End Win32 API output ----
```

Let me know if you need further details

### PR DESCRIPTION
**Pull request by carpnick**
...up for windows, need to use rights instead.

Fixes #310

### CODE REVIEW

**Comment by carpnick** [comment]
Build is failing due to OHAI issue.  Should be unrelated:
An error occurred while installing ohai (8.0.1), and Bundler cannot continue.

**Comment by carpnick** [comment]
Related: https://github.com/kohsuke/winsw/pull/64
Needed to make it fully automated outside of a domain

**Comment by isuftin** [comment]
@carpnick I'm seeing the same ohai issue from my PR 10 days ago

**Comment by loky9000** [comment]
not working for LocalSystem user
directory("C:\\jenkins") do
  provider Chef::Provider::Directory
  action :create
  retries 0
  retry_delay 2
  guard_interpreter :default
  path "C:\\\\jenkins"
  recursive true
  rights [{:permissions=>:full_control, :principals=>"LocalSystem"}]
end


================================================================================
Error executing action `create` on resource 'jenkins_windows_slave[10.73.45.204]'
================================================================================

Chef::Exceptions::Win32APIError
-------------------------------
directory[C:\\jenkins] (dynamically defined) had an error: Chef::Exceptions::Win32APIError: No mapping between account names and security IDs was done.
---- Begin Win32 API output ----
System Error Code: 1332
System Error Message: No mapping between account names and security IDs was done.
---- End Win32 API output ----

**Comment by carpnick** [comment]
Hey @loky9000 ,

Thanks for the review. This is a chef platform resource "directory".  I have never tried to use localsystem/localservice/networkservice accounts on that resource.

Have you been able to get that to work in a typical recipe?

**Comment by loky9000** [comment]
No i haven't.
LocalSystem is default user for  jenkins_windows_slave resource. 
in my recipe i use 
jenkins_windows_slave node['ipaddress'] do
      remote_fs 'C:\jenkins'
    end

it work fine for user 'Administrator' but in this case service jenkins slave unable to start.


================================================================================
Error executing action `create` on resource 'jenkins_windows_slave[10.73.90.8]'
================================================================================

SystemCallError
---------------
service[jenkins-slave] (dynamically defined) had an error: SystemCallError: The service did not start due to a logon failure. - StartService: The service did not start due to a logon failure.

Resource Declaration:
---------------------
# In C:/cygwin64/tmp/tmp.0amTO7ZKOR/cookbooks/cookbook_qubell_jenkins/recipes/node.rb

 38:     jenkins_windows_slave node['ipaddress'] do
 39:       remote_fs 'C:\jenkins'
 40:       user      'Administrator'
 41:     end
 42:   end

Compiled Resource:
------------------
# Declared in C:/cygwin64/tmp/tmp.0amTO7ZKOR/cookbooks/cookbook_qubell_jenkins/recipes/node.rb:38:in `from_file'

jenkins_windows_slave("10.73.90.8") do
  action :create
  retries 0
  retry_delay 2
  guard_interpreter :default
  cookbook_name :cookbook_qubell_jenkins
  recipe_name "node"
  remote_fs "C:\\\\jenkins"
  user "Administrator"
  slave_name "10.73.90.8"
  usage_mode "normal"
  in_demand_delay 0
  idle_delay 1
  description "Jenkins slave 10.73.90.8"
  executors 1
  service_name "jenkins-slave"
  winsw_url "http://repo.jenkins-ci.org/releases/com/sun/winsw/winsw/1.16/winsw-1.16-bin.exe"
  winsw_checksum "052f82c167fbe68a4025bcebc19fff5f11b43576a2ec62b0415432832fa2272d"
end

**Comment by carpnick** [comment]
Ok this could be caused by a couple issues:

Problem 1:
If the account you are using is not in the security group "Allow logon as a local service", you will see the error you have seen.

Solution:
This is WinSW related.  The new WinSW 1.17 has my PR below integrated.  I will need to do a second PR here in Jenkins cookbook to modify the config to automatically add any user account specified in the config to the "Allow logon as a local service" policy as an optional item on the LWRP call.  For now this process can only be done via C++ apis, thus it was integrated into WinSW.
Related: kohsuke/winsw#64

Problem 2:
I see you are not passing in a password.  Might try something like the following:

Solution: When the service starts it is trying to start the process as Administrator.  It needs a valid password:
jenkins_windows_slave 'mywinslave' do
  remote_fs 'C:/jenkins'
  user       'Administrator'
  password   'MyPassword'
end

**Comment by loky9000** [comment]
Thanks!
I'll try today both solutions, but I think 
directory creation need to be fixed for LocalSystem because this is a default user.

File: libraries/slave_jnlp.rb [hunk 1]
```diff
@@ -54,19 +54,18 @@ def load_current_resource
 
     def action_create
       super
-
-      parent_remote_fs_dir_resource.run_action(:create)
-
-      # don't create user/group on Windows
+
```
Comment (ksubrama) [inline libraries/slave_jnlp.rb]: I'm curious - why did you move this code into the unless block.  The overriden code below seems to do the same thing as the one here.  Wouldn't just overriding remote_fs_dir_resource be enough?

File: libraries/slave_windows.rb [hunk 2]
```diff
@@ -81,9 +80,24 @@ def action_create
...
     # Embedded Resources
 
+    # Creates a `directory` resource that represents the directory
+    # specified the `remote_fs` attribute. The caller will need to call
+    # `run_action` on the resource.
+    #
+    # @return [Chef::Resource::Directory]
+    #
+    def remote_fs_dir_resource
+      return @remote_fs_dir_resource if @remote_fs_dir_resource
```
Comment (ksubrama) [inline libraries/slave_windows.rb]: This duplicates code from the parent class.  Could you put the offending "owner/group" statements in an "unless windows" clause - or even better, pull those out to a "remote_fs_dir_resource_perms" method.  This way the parent can do it's unixy permissions thing and you'd need to override just that method here to do the windows dance.

File: libraries/slave_windows.rb [hunk 3]
```diff
@@ -128,6 +142,22 @@ def slave_compat_xml
       @slave_compat_xml
     end
 
+    
+    def user_hash
```
Comment (ksubrama) [inline libraries/slave_windows.rb]: Man, we really should have a common utility method in Chef core to do that.

**Comment by ksubrama** [comment]
Thanks for this fix!  I have some minor aesthetics comments.  Otherwise it looks good to me (modulo comments from someone who knows this better than me).  If you can rebase ahead, I'd be happy to merge it in.

File: libraries/slave_jnlp.rb (hunk 1 above)
Comment (carpnick) [inline libraries/slave_jnlp.rb]: @ksubrama Thought it was easier to understand what was going on.  Usually if I see a bunch of conditional statements in a class, and we already have a windows specific class,  thought it would make it easier to understand.

File: libraries/slave_windows.rb (hunk 2 above)
Comment (carpnick) [inline libraries/slave_windows.rb]: @ksubrama  I think that might be a better idea.  Let me know if you agree with the reasoning above, or if we should just make this change alot smaller and override the perms method within the windows class and call it a day.

As stated above, thought the code was a bit more straight forward when separating the constructor logic for unix/windows.

File: libraries/slave_windows.rb (hunk 3 above)
Comment (carpnick) [inline libraries/slave_windows.rb]: agreed.  Problem is this is [WinSW] ( https://github.com/kohsuke/winsw) related which  might have different requierments.  But with that said, the regex and split I would say makes sense to be common.

File: libraries/slave_jnlp.rb (hunk 1 above)
Comment (ksubrama) [inline libraries/slave_jnlp.rb]: Fair enough.  Looks good to me.

**Comment by ksubrama** [comment]
:+1:

**Comment by carpnick** [comment]
Closing in favor of #333
